RD-1758 / RD-1761 - Add support for type: 'attribute' and operators: 'any_of' and 'not_any_of'

qooban opened the pull request:
### Changes
* Added support for type `attribute` and operators: `any_of` and `not_any_of` which means that there is multiple values dropdown for Blueprint, Site Name and Creator attributes
* Fixed support for type `label` by putting label key in the right place (in `key` field in filter rule, instead of first value in `values field)
* Remove wrapping `Stage.Common.Filters.Form` content inside `Form` component to allow using `Form` outside of the `Stage.Common.Filters.Form`

### Video
https://user-images.githubusercontent.com/5202105/114385370-3f07be00-9b90-11eb-864a-2f4eda867f42.mp4

qooban commented on widgets/common/src/DynamicDropdown.jsx:
```diff
@@ -3,25 +3,25 @@ import './DynamicDropdown.css';
 
 let instanceCount = 0;
 
-function DynamicDropdown({
```
I had to add props' default values to use `DynamicDropdown` in TS files and not specify all default props.

(at least that is my understanding why I had to do so 🙂 )

qooban commented:
I'll run stage system tests once RD-2026 is fixed.

kubama commented on widgets/common/src/filters/inputs/BlueprintValueInput.tsx:
```diff
@@ -0,0 +1,16 @@
+/* eslint-disable react/jsx-props-no-spreading */
```
Isn't it disabled globally already?

qooban replied:
Unfortunately. RD-1584 is still open.

Gelio commented on widgets/common/src/filters/inputs/LabelValueInput.tsx:
```diff
@@ -0,0 +1,43 @@
+        // TODO(RD-2007): Add better styling
+        <>
+            <LabelKeyDropdown innerRef={keyDropdownRef} onChange={onKeyChange} toolbox={toolbox} value={labelKey} />
+            {(operator === LabelsFilterRuleOperators.AnyOf || operator === LabelsFilterRuleOperators.NotAnyOf) && (
```
Optional:

A useful pattern for such matches would be:

```suggestion
            {operatorsWithValues.includes(operator) && (
```

and having an `operatorsWithValues` array that contains those 2 operators :slightly_smiling_face: 

Both approaches are fine, I'm leaving that for you to consider :smile:

qooban replied:
I like that idea.
Please correct me if I'm wrong, but I see I had to define `operatorsWithValues` the following way:
```
const operatorsWithValues: FilterRuleOperator[] = [LabelsFilterRuleOperators.AnyOf, LabelsFilterRuleOperators.NotAnyOf];
```
to be able to use it in
```
operatorsWithValues.includes(operator)
```

I assume that's the right way or there is simpler way to do it?

Gelio replied:
If TS is not happy with omitting the type annotation of `FilterRuleOperator[]`, then yes, that should be the way :slightly_smiling_face:

qooban replied:
Yes, TS was complaining about it. OK, done.

Gelio commented on widgets/common/src/filters/inputs/AttributeValueInput.tsx:
```diff
@@ -0,0 +1,46 @@
+import { FilterRuleOperator, LabelsFilterRuleOperators } from '../types';
+import DynamicDropdown from '../../DynamicDropdown';
+
+interface AttributeValueInputProps {
```
I reckon we could extend `CommonAttributeValueInputProps` here :slightly_smiling_face: That could save us a few lines of code

qooban replied:
Yeah, of course. That was actually my idea as well, but I forgot about it during rush development 🙂 Thanks, I'll change that.

Gelio reviewed (commented):
I like the clear folder structure and the reuse of `AttributeValueInput` :+1:

qooban commented:
@Gelio @kubama Thanks for your reviews. I think I'll manage to address all of the comments tomorrow morning and ask you then for the final review.

qooban commented on widgets/common/src/DynamicDropdown.jsx:
```diff
@@ -3,25 +3,25 @@ import './DynamicDropdown.css';
-    disabled,
-    multiple,
-    placeholder,
+export default function DynamicDropdown({
```
The main reason why I did not remove it is that we would have a lot of complains from ESLint:

> ESLint: propType "disabled" is not required, but has no corresponding defaultProps declaration.(react/require-default-props)

Gelio replied:
IMO an `eslint-disable` comment for that rule is more suitable and safer than having the default values defined in 2 places, which causes confusion on my side :slightly_smiling_face: I still suggest removing the default props

qooban replied:
OK, disabled the rule.

qooban commented on widgets/common/src/filters/FiltersDefinitionForm.tsx:
```diff
@@ -64,7 +63,7 @@ const FiltersDefinitionForm: FunctionComponent<FiltersDefinitionFormProps> = ({
     }, [rows]);
 
     return (
-        <Form>
```
Yes, there was ask from @kubama to have it like that, and I think it make sense.

`FiltersDefinitionForm` component will be used as part of other `Form` and to avoid having nested `Form` I decided to remove the wrapper.

If there is a real use case for having `Form` wrapper here we can always add a prop. Do you see such use case right now?

Gelio replied:
No, I don't :smile: I was asking since this code change IMO is not required in PR, as there is no other code that uses it :slightly_smiling_face: To my mind, it would be less confusing to have it removed in the PR that actually leverages the functionality and embeds the component in some other form :slightly_smiling_face:

Also, I now see another problem - a naming one. Does that not look confusing?

```tsx
<Form>
  <FiltersDefinitionForm />
</Form>
```

In my view, it does, so optionally, I encourage changing the `FiltersDefinitionForm` component name to `FilterRuleRows` (or some other name, but one that does not suggest this is a fully form) :slightly_smiling_face:

qooban replied:
> I was asking since this code change IMO is not required in PR, as there is no other code that uses it 
I could create a separate PR, but I don't think it is worth to do so in such "rushy" times we have right now.

I got your point, but I'm not sure is your suggestion the right one (`FilterRuleRows`).
It's not only rule rows, there is also `Add new rule` button, which for me is making it to be a form.

From the outside of `common/filters` it's more probable that we will use `Stage.Common.Filters.Form` not `FiltersDefinitionForm`.

We actually did the same (provided `*Form` component without `Form` wrapper) in `UploadPluginForm`. I think it is not big sin to do so.

Gelio replied:
Hmm... Ok, you're right, this is a minor problem :slightly_smiling_face: Let's not waste our time

As for

> I could create a separate PR, but I don't think it is worth to do so in such "rushy" times we have right now.

In my mind I imagined that @kubama could have removed the `Form` wrapper in his PR, not you creating a separate PR :slightly_smiling_face: Then, the change would be made in the same PR (and eventually, the same commit on `master`) that takes advantage of the change, resulting in a healthier history :slightly_smiling_face: That's probably worth keeping in mind in the future (I got a similar comment some time ago and I thought it was valuable feedback :smile: )

qooban replied:
Make sense 👍🏻

qooban commented on widgets/common/src/filters/consts.ts:
```diff
@@ -1,2 +1,3 @@
 // eslint-disable-next-line import/prefer-default-export
+export const { i18n } = Stage;
```
🙂  Do you see any issues with it? 

My reasoning was this: whenever we use `i18nPrefix` we also need `i18n.t`, so it is one line less in each file to get `i18n` extracted from `Stage`.

Gelio replied:
Nope, no problem here, just a comment about a novel way of exporting :slightly_smiling_face:

Gelio replied:
Perhaps it would make sense to export `i18n` as a global variable then, instead of exposing it on `Stage`. Then, we would not have to unwrap it from `Stage` each time we want to use it, and just use `i18n.t` whereever

qooban commented on widgets/common/src/filters/consts.ts:
```diff
@@ -1,2 +1,3 @@
 // eslint-disable-next-line import/prefer-default-export
```
Right, I'll remove it.

qooban commented on widgets/common/src/filters/RuleValueInput.tsx:
```diff
@@ -1,63 +1,47 @@
-        );
+            );
+        default:
+            return null;
```
Yep, I'll change it.

qooban commented on widgets/common/src/filters/inputs/AttributeValueInput.tsx:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    // TODO(RD-1762): Add support for type 'attribute' and operators different from 'any_of' and 'not_any_of'
+    const textValue = _.castArray(value)[0] || '';
```
That's temporary, but I'll change it.

qooban commented on widgets/common/src/filters/inputs/BlueprintValueInput.tsx:
```diff
@@ -0,0 +1,16 @@
+const BlueprintValueInput: FunctionComponent<CommonAttributeValueInputProps> = props => (
+    <AttributeValueInput
+        fetchUrl="/blueprints?state=uploaded&_include=id"
+        placeholder={i18n.t(`${i18nPrefix}.inputsPlaceholders.blueprintId`)}
```
Sure, will do.

qooban commented on widgets/common/src/filters/inputs/LabelValueInput.tsx:
```diff
@@ -0,0 +1,43 @@
+import { useRef } from 'react';
+
+import { LabelsFilterRuleOperators } from '../types';
+import LabelKeyDropdown from '../../labels/KeyDropdown';
+import LabelValueDropdown from '../../labels/ValueDropdown';
```
👍🏻

qooban commented:
I hope I answered/addressed all the comments, so I'm waiting for final reviews.

As our system test CI looks healthy now, I started: https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/435/

Gelio commented on widgets/common/src/filters/consts.ts:
```diff
@@ -1,3 +1,3 @@
-// eslint-disable-next-line import/prefer-default-export
 export const { i18n } = Stage;
 export const i18nPrefix = 'widgets.common.filters.form';
+export const i18nPlaceholdersPrefix = 'widgets.common.filters.form.inputsPlaceholders';
```
Optional:

We could reference the `i18nPrefix` here

```suggestion
export const i18nPlaceholdersPrefix = `${i18nPrefix}.inputsPlaceholders`;
```

qooban replied:
Right, fixed.

Gelio commented on widgets/common/src/filters/inputs/AttributeValueInput.tsx:
```diff
@@ -1,15 +1,14 @@
-import { FilterRuleOperator, LabelsFilterRuleOperators } from '../types';
+
 import DynamicDropdown from '../../DynamicDropdown';
+import { LabelsFilterRuleOperators } from '../types';
+import { CommonAttributeValueInputProps } from './types';
```
I'm guessing the 2nd import was supposed to be an `import type` :smile: 

Optional:

Let's add the `import type` or merge the 2 imports into one :slightly_smiling_face:

qooban replied:
Yeah, I would like WebStorm do that for me 🙂 (maybe I have to configure sth)
Fixed.

Gelio replied:
@kubama also mentioned it would be nice to have it done automatically :slightly_smiling_face: Not sure if it can be automatic, but it can definitely be detected as an error (see RD-2000) :smile:

kubama replied:
I didn't check, but hope the rule would be auto-fixable. If yes, it would aid the WebStorm type auto-import issue.

Gelio replied:
Nn update here - in my private project I have enabled `importsNotUsedAsValues` https://github.com/Gelio/supreme-trobot/blob/main/tsconfig.json#L26 and vim (that uses the TypeScript Language Service, that VSCode and [probably WebStorm also use](https://www.jetbrains.com/help/webstorm/typescript-support.html#ws_verify_typescript):

> WebStorm verifies TypeScript code mainly based on the data from the TypeScript Language Service which also compiles TypeScript into JavaScript.

) and when I use the auto-import feature, `import type` is used automatically :smile: 

Yet another reason to complete RD-2000 :smile:

Gelio reviewed (dismissed):
LGTM :+1: Minor comments left :smile:

qooban commented:
Thanks for the reviews.

All passed in https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/435 so I'll merge this PR once Jenkins (Stage Multipbranch pipeline) gives me approval as well 🙂
